File: uwp/ArtistWayUWP/Services/SessionService.cs
using System;
using Windows.Data.Json;
using Windows.Security.Credentials;

namespace ArtistWayUWP.Services
{
    public sealed class FirebaseSession
    {
        public string Uid { get; set; }
        public string IdToken { get; set; }
        public string RefreshToken { get; set; }
        public string Provider { get; set; }
        public string Email { get; set; }
        public string DisplayName { get; set; }
        public DateTimeOffset IdTokenExpiresAt { get; set; }

        // Um pouco de folga antes do vencimento de verdade, pra nunca correr
        // o risco de o Firestore rejeitar por token vencido no meio de uma
        // chamada em andamento.
        public bool NeedsRefresh => DateTimeOffset.UtcNow >= IdTokenExpiresAt.AddSeconds(-60);
    }

    // Guarda a sessão do login (uid/tokens do Firebase) no cofre de
    // credenciais do Windows (PasswordVault) — criptografado pelo sistema,
    // diferente do resto dos dados do app (JSON simples em LocalDataStore),
    // porque um refresh token equivale a uma senha.
    public static class SessionService
    {
        private const string Resource = "ArtistWayCompanion.FirebaseSession";
        private const string UserName = "session";

        public static void SaveSession(AuthResult result)
        {
            PasswordVault vault = new PasswordVault();
            RemoveExisting(vault);

            DateTimeOffset expiresAt = DateTimeOffset.UtcNow.AddSeconds(result.ExpiresInSeconds > 0 ? result.ExpiresInSeconds : 3600);
            JsonObject json = new JsonObject
            {
                ["uid"] = JsonValue.CreateStringValue(result.FirebaseUid ?? ""),
                ["idToken"] = JsonValue.CreateStringValue(result.FirebaseIdToken ?? ""),
                ["refreshToken"] = JsonValue.CreateStringValue(result.FirebaseRefreshToken ?? ""),
                ["provider"] = JsonValue.CreateStringValue(result.Provider ?? ""),
                ["email"] = JsonValue.CreateStringValue(result.FirebaseEmail ?? ""),
                ["displayName"] = JsonValue.CreateStringValue(result.FirebaseDisplayName ?? ""),
                ["idTokenExpiresAt"] = JsonValue.CreateStringValue(expiresAt.ToString("o")),
            };

            vault.Add(new PasswordCredential(Resource, UserName, json.Stringify()));
        }

        // Chamado pelo SyncService depois de renovar o idToken via refresh
        // token (o refresh token em si não muda nesse fluxo do Firebase).
        public static void UpdateTokens(string idToken, int expiresInSeconds)
        {
            FirebaseSession session = GetSession();
            if (session == null)
            {
                return;
            }

            PasswordVault vault = new PasswordVault();
            RemoveExisting(vault);

            JsonObject json = new JsonObject
            {
                ["uid"] = JsonValue.CreateStringValue(session.Uid ?? ""),
                ["idToken"] = JsonValue.CreateStringValue(idToken ?? ""),
                ["refreshToken"] = JsonValue.CreateStringValue(session.RefreshToken ?? ""),
                ["provider"] = JsonValue.CreateStringValue(session.Provider ?? ""),
                ["email"] = JsonValue.CreateStringValue(session.Email ?? ""),
                ["displayName"] = JsonValue.CreateStringValue(session.DisplayName ?? ""),
                ["idTokenExpiresAt"] = JsonValue.CreateStringValue(DateTimeOffset.UtcNow.AddSeconds(expiresInSeconds > 0 ? expiresInSeconds : 3600).ToString("o")),
            };

            vault.Add(new PasswordCredential(Resource, UserName, json.Stringify()));
        }

        public static FirebaseSession GetSession()
        {
            try
            {
                PasswordVault vault = new PasswordVault();
                PasswordCredential credential = vault.Retrieve(Resource, UserName);
                credential.RetrievePassword();
                JsonObject json = JsonObject.Parse(credential.Password);
                return new FirebaseSession
                {
                    Uid = json.GetNamedString("uid", ""),
                    IdToken = json.GetNamedString("idToken", ""),
                    RefreshToken = json.GetNamedString("refreshToken", ""),
                    Provider = json.GetNamedString("provider", ""),
                    Email = json.GetNamedString("email", ""),
                    DisplayName = json.GetNamedString("displayName", ""),
                    IdTokenExpiresAt = DateTimeOffset.TryParse(json.GetNamedString("idTokenExpiresAt", ""), out DateTimeOffset expiresAt)
                        ? expiresAt
                        : DateTimeOffset.UtcNow.AddSeconds(-1),
                };
            }
            catch (Exception)
            {
                return null;
            }
        }

        public static void ClearSession()
        {
            RemoveExisting(new PasswordVault());
        }

        private static void RemoveExisting(PasswordVault vault)
        {
            try
            {
                vault.Remove(vault.Retrieve(Resource, UserName));
            }
            catch (Exception)
            {
                // não tinha sessão salva — nada a limpar.
            }
        }
    }
}

File: android/ArtistWayAndroid/app/src/main/java/com/rodcarvalho/artistway/auth/SessionService.kt
package com.rodcarvalho.artistway.auth

import android.content.Context
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import java.time.Instant

@Serializable
data class FirebaseSession(
    val uid: String,
    val idToken: String,
    val refreshToken: String,
    val provider: String,
    val email: String,
    val displayName: String,
    val idTokenExpiresAt: String,
) {
    // Um pouco de folga antes do vencimento de verdade, pra nunca correr
    // o risco do Firestore rejeitar por token vencido no meio de uma
    // chamada em andamento.
    fun needsRefresh(): Boolean {
        val expiresAt = runCatching { Instant.parse(idTokenExpiresAt) }.getOrNull() ?: return true
        return !Instant.now().isBefore(expiresAt.minusSeconds(60))
    }
}

// Guarda a sessão do login (uid/tokens do Firebase) em
// EncryptedSharedPreferences — criptografado pelo Android Keystore,
// diferente do resto dos dados do app (JSON simples em LocalDataStore),
// porque um refresh token equivale a uma senha. Mesmo papel do
// PasswordVault no UWP.
object SessionService {
    private const val PREFS_NAME = "artistway_session"
    private const val KEY_SESSION = "session"

    private lateinit var appContext: Context
    private val json = Json { ignoreUnknownKeys = true }

    fun init(context: Context) {
        appContext = context.applicationContext
    }

    private fun prefs() = EncryptedSharedPreferences.create(
        appContext,
        PREFS_NAME,
        MasterKey.Builder(appContext).setKeyScheme(MasterKey.KeyScheme.AES256_GCM).build(),
        EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
        EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM,
    )

    fun saveSession(session: FirebaseSession) {
        prefs().edit().putString(KEY_SESSION, json.encodeToString(FirebaseSession.serializer(), session)).apply()
    }

    // Chamado pelo SyncService depois de renovar o idToken via refresh
    // token (o refresh token em si não muda nesse fluxo do Firebase).
    fun updateTokens(idToken: String, expiresInSeconds: Int) {
        val current = getSession() ?: return
        val expiresAt = Instant.now().plusSeconds(if (expiresInSeconds > 0) expiresInSeconds.toLong() else 3600L)
        saveSession(current.copy(idToken = idToken, idTokenExpiresAt = expiresAt.toString()))
    }

    fun getSession(): FirebaseSession? {
        val text = prefs().getString(KEY_SESSION, null) ?: return null
        return runCatching { json.decodeFromString(FirebaseSession.serializer(), text) }.getOrNull()
    }

    fun clearSession() {
        prefs().edit().remove(KEY_SESSION).apply()
    }
}
